### PR TITLE
Remove some tech debt from web/query (I think)

### PR DESCRIPTION
**Pull request by fpacifici**
This is the refactoring I promised last week during the reviews of #840.

In practice this does not add any new abstraction nor it makes changes to the query processing architecture. The goal is to redistribute responsibilities in web/query in a way that makes more sense in the new query processing pipeline. This is just a first step. There is a lot more to do.

In detail:
- splits what is the execution of a DB query (still not necessarily clickhosue specific) from the entire query processing process. This goes in db_query.py and does not depend on Dataset.
- reorder the methods in web/query to follow the process. I thought the previous organization was very counterintuitive.
- reduce dependencies on dataset.
- provides some docs on the responsibilities.

### CODE REVIEW

File: snuba/web/db_query.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    stats.update(query_settings)
+
+    query_metadata.query_list.append(
+        ClickhouseQueryMetadata(sql=sql, stats=stats, status=status, trace_id=trace_id)
+    )
+
+    return stats
+
+
+def raw_query(
```
Comment (fpacifici) [inline snuba/web/db_query.py]: I made almost no change to this method. Just copied over to a separate file and removed the dataset dependency.

File: snuba/web/query.py [hunk 2]
```diff
@@ -280,70 +57,29 @@ def parse_and_run_query(
...
-        if scope.span:
-            if "max_threads" in stats:
-                scope.span.set_tag("max_threads", stats["max_threads"])
-
-    return result
-
-
 @split_query
-def _process_and_run_query(
+def _run_query_pipeline(
```
Comment (fpacifici) [inline snuba/web/query.py]: This methods now owns the query processing pipeline, and only that.

File: snuba/web/query.py [hunk 3]
```diff
@@ -2,263 +2,35 @@
...
-    stats.update(query_settings)
-
-    query_metadata.query_list.append(
-        ClickhouseQueryMetadata(sql=sql, stats=stats, status=status, trace_id=trace_id)
-    )
-
-    return stats
-
-
-def record_query(
```
Comment (fpacifici) [inline snuba/web/query.py]: This is just at the bottom of the file now.

File: snuba/web/query.py [hunk 4]
```diff
@@ -2,263 +2,35 @@
...
-            tags={
-                "status": query_metadata.status,
-                "referrer": referrer,
-                "final": final,
-            },
-            mark_tags={"final": final},
-        )
-
-
 def parse_and_run_query(
```
Comment (fpacifici) [inline snuba/web/query.py]: This is the entry point now. Did not touch that not to break everything else in the codebase. 
But, after moving split query in the Query Execution Strategy for Storage split, this method will not need to exist anymore.

File: snuba/web/query.py [hunk 5]
```diff
@@ -268,8 +40,13 @@ def parse_and_run_query(
         query_list=[],
     )
 
+    with sentry_sdk.configure_scope() as scope:
```
Comment (fpacifici) [inline snuba/web/query.py]: There was no need to add these tags (transaction tags) later in the flow.

**Review by tkaemming — APPROVED** [review]
This is a big improvement, thanks for taking the time to do this.

File: snuba/web/db_query.py (hunk 1 above)
Comment (fpacifici) [reply snuba/web/db_query.py]: Sorry my comment was not accurate. I wanted to keep what depends on dataset in a file and what does not in another. There was no dependency from this specific  function.

File: snuba/web/__init__.py [hunk 6]
```diff
@@ -1,8 +1,19 @@
...
 
+class RawQueryException(Exception):
+    def __init__(
+        self, err_type: str, message: str, stats: Mapping[str, Any], sql: str, **meta
+    ):
+        self.err_type = err_type
+        self.message = message
+        self.stats = stats
+        self.sql = sql
+        self.meta = meta
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/web/__init__.py]: It makes sense to me, at least as long as we have clickhosue only.

**Pull request merged**